_posts/2021-12-29-readable-tests-by-example.md

```diff
@@ -0,0 +1,552 @@
+---
+layout: post
+title: "Make your tests readable by example"
+author: [kamil.jedrzejuk]
+tags: [tech, backend, testing, tdd, bdd, java, groovy, ddd]
+---
+
+Have you ever worked on a project where after downloading the code from the repository you start to wonder what business
+requirements are hidden under the layer of unreadable tests? \
+Or maybe you are currently wondering how to test a new feature that you have been entrusted to implement?
+
+Nothing is more frustrating than code that we cannot control, and over time it becomes so troublesome that no one
+bothers about it too much. People do not pay enough attention to such a project and subsequent tests, if any, duplicate
+the shortcomings of all the previous ones in the legacy.
+
+In this article I will show you how to disenchant this miserable state of affairs with a few simple tips. I invite
+you to approach the tests in a completely different way to create their new version.
+
+## Hello domain!
+
+The domain that will serve us as a background will not be too complicated, but at the same time not simple enough to
+actually be able to capture the benefit for which it is worth investing more time and effort in writing tests that are
+clear and easy to develop further.
+
+For the purpose of presenting to you a domain that is neither too complicated nor too trivial, I will use a model of a
+vinyl records online shop (I tried to create a domain which is universal and intuitive at the same time).
+
+Let’s assume that we have the following very preliminary and explicitly written general business assumptions.
+
+```text
+An online store sells vinyl records. Each order is delivered by a courier company cooperating with the store.
+
+The cost of delivery is charged when the customer pays for the order.
+
+The cost of delivery is always collected from the supplier’s system (the courier’s system).
+
+In the event of its unavailability (e.g. when the external courier system cannot provide the cost amount),
+we can assume that the cost of delivery is always a fixed amount of EUR 20.
+
+We distinguish between two types of clients: STANDARD and VIP.
+
+If the order is processed for a customer with a VIP status or the value of the order exceeds a certain amount
+according to the running promotional campaign (current price list configuration), the order will be delivered free
+of charge.
+
+Additionally, for the VIP customer, a free music track should be sent to their mailbox after the payment of the order.
+
+After paying for the order, no modifications can be made.
+```
+
+Such assumptions can be translated into one of the BDD-style scenarios:
+<img alt="API package structure" src="/img/articles/2021-12-29-readable-tests-by-example/1.png" title="Scenario 1.1">
+Scenario 1.1
```

Comment:
Zastanawiam się czy musimy to mieć jako obrazek no i czy sprawdzałeś może jak ten tekst `Scenario 1.1` wygląda na wyrenderowanej stronce?

Reply:
Tak w ogóle to najlepiej skorzystać z markdowna https://kramdown.gettalong.org/syntax.html#images

Reply:
Jeśli chodzi o zdjęcie wygląda to tak:
![image](https://user-images.githubusercontent.com/10248783/152152850-5ddd5820-b297-4365-a999-bae9d7b0e26b.png)

Reply:
Albo przepisujac na gherkina. 
![image](https://user-images.githubusercontent.com/10248783/152153979-b15e7dfa-07af-438b-82d0-152ceb330291.png)
```
Such assumptions can be translated into one of the BDD-style scenarios:

```gherkin
    Scenario: 1.1
        Given There is an unpaid order
        And The Client is a VIP
        When The client makes the payment
        Then Payment succeeded
        And The client did not paid for delivery
        And Free music track was sent to the client
```

```

Reply:
No to pewnie najlepiej jako code

Reply:
@wpanas  Done.

Reply:
Nie widzę odpowiedzi więc resolved.